### PR TITLE
Always onnext

### PR DESCRIPTION
**Pull request by steveorsomethin**
Currently, a get() for paths that are missing in the cache behaves differently than a get() for atoms of undefined. This PR makes both conditions produce identical results, and ensures onNext() is always called on the model result.

### CODE REVIEW

File: test/falcor/get/get.dataSource-and-cache.spec.js [hunk 1]
```diff
@@ -582,7 +588,7 @@ describe('DataSource and Partial Cache', function() {
                         done();
                     },
                     function() {
-                        expect(onNextSpy.callCount).to.equal(0);
+                        expect(onNextSpy.callCount).to.equal(1);
```
Comment (jhusain) [inline test/falcor/get/get.dataSource-and-cache.spec.js]: Can we confirm the contents of the body is an empty object?

**Comment by lrowe** [comment]
Testing locally I see that we're not onNexting when we request no paths:
```
$ node
> var falcor = require("./lib")
undefined
> var model = new falcor.Model({cache: {foo: 1}})
undefined
> var sub = model.get().subscribe(console.log)
undefined
```

**Comment by steveorsomethin** [comment]
Oops, good catch. Doesn't seem like someone would _actually_ want to request nothing, but I see a consistency argument. Will fix.

**Comment by steveorsomethin** [comment]
Let me dig in a little more before you review further. The empty get thing exposed some additional problems.

**Comment by steveorsomethin** [comment]
Ok, proceed.

**Comment by lrowe** [comment]
Another minor weirdness:
```
$ node
> var falcor = require("./lib")
undefined
> var model = new falcor.Model({cache: {foo: 1}})
undefined
> var s = model.asDataSource().get(['a']).subscribe(console.log)
{ jsonGraph: { a: { '$type': 'atom' } }, paths: [ [ 'a' ] ] }
undefined
> var s = model.asDataSource().get([]).subscribe(console.log)
{ jsonGraph: {} }
undefined
```
I would expect the second JSONGraph response to include an empty paths array.

**Comment by steveorsomethin** [comment]
Thanks, fixed + test.

File: test/falcor/set/set.dataSource-and-cache.spec.js [hunk 2]
```diff
@@ -159,6 +159,64 @@ describe('DataSource and Cache', function() {
...
+                            genreList: {
+                                0: {
+                                    0: {
+                                        summary: 1337
+                                    }
+                                }
+                            }
+                        }
+                    });
+                }).
```
Comment (jhusain) [inline test/falcor/set/set.dataSource-and-cache.spec.js]: Please assert against the value of the second onNext.

File: test/get-core/values.spec.js [hunk 3]
```diff
@@ -187,10 +187,57 @@ describe('Values', function() {
...
+                videos: {
+                    123: {
+                        name: atom()
+                    }
+                }
+            }
+        });
+    });
+
     it('should have no output for empty paths.', function() {
```
Comment (jhusain) [inline test/get-core/values.spec.js]: This test is probably not named appropriately given we are outputting data for empty paths.

File: test/get-core/values.spec.js [hunk 4]
```diff
@@ -187,10 +187,57 @@ describe('Values', function() {
...
+                            
+                        }
+                    }
+                }
+            },
+            cache: cacheGenerator(0, 1)
+        });
+    });
+
+    it('should have no output for empty get.', function() {
```
Comment (jhusain) [inline test/get-core/values.spec.js]: Test not named appropriately given we're outputting data.

File: test/integration/call.spec.js [hunk 5]
```diff
@@ -113,7 +113,7 @@ describe('call', function() {
             call("genreList[0].titles.push", args)).
             doAction(onNext, noOp, noOp).
             subscribe(noOp, done, function() {
-                expect(onNext.callCount).to.equal(0);
+                expect(onNext.callCount).to.equal(1);
```
Comment (jhusain) [inline test/integration/call.spec.js]: Can we assert on onNext contents?

File: lib/response/get/checkCacheAndReport.js [hunk 6]
```diff
@@ -17,33 +17,36 @@ var getWithPathsAsPathMap = gets.getWithPathsAsPathMap;
  * @private
  */
 module.exports = function checkCacheAndReport(model, requestedPaths, observer,
```
Comment (jhusain) [inline lib/response/get/checkCacheAndReport.js]: Can you document the fields in seed?

**Comment by lrowe** [comment]
## Re: should not onNext when only receiving errors / should onNext when receiving errors and missing paths (or values)

Are we being consistent between progressive and non-progressive mode here?

### Without treatErrorsAsValues

PBE's effectively become critical errors since view code does not want to account for them.

#### Non-progressive mode without treatErrorsAsValues

We only want to render the final view.

I don't see the purpose in onNext'ing any value when it will be followed by an onError. For this case I think developers usually want behaviour more like a Promise / Rx Single / async Result, so either onNext exactly once or onError but never both.

#### Progressive mode without treatErrorsAsValues

We want the opportunity to display a partial view before rendering the final view. If there are any errors then we render an error view instead.

What's the expected behaviour in progressive mode when:

1. No errors in cache and a mix of values and errors are returned by the datasource

    I would expect an immediate (possibly empty) onNext followed by an onError when the datasource returns even if there are values.

2. All errors in cache, no unknown values

    No need to make a datasource request, onError immediately.

3. Some errors in cache, some unknown values

    Do we even need to make a datasource request here? If we're going to render an error view we may as well display it immediately.

### With treatErrorsAsValues / boxedValues

In this case PBEs in the cache are treated as values so do not trigger an error. Only critical errors would result in same behaviour as above.

**Pull request merged**